utilities/transactions/write_prepared_transaction_test.cc

```diff
@@ -758,6 +778,129 @@ TEST_P(WritePreparedTransactionTest, IsInSnapshotTest) {
   }
 }
 
+TEST_P(WritePreparedTransactionTest, CompactionShouldKeepUncommittedKeys) {
+  // Snapshots to avoid keys get evicted.
+  std::vector<const Snapshot*> snapshots;
+  // Each key here represent a standalone test case.
+  ASSERT_OK(db->Put(WriteOptions(), "key1", "value1_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Put(WriteOptions(), "key2", "value2_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Put(WriteOptions(), "key3", "value3_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Put(WriteOptions(), "key4", "value4_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Merge(WriteOptions(), "key5", "value5_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Merge(WriteOptions(), "key5", "value5_2"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Put(WriteOptions(), "key6", "value6_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Put(WriteOptions(), "key7", "value7_1"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->Flush(FlushOptions()));
+  ASSERT_OK(db->Delete(WriteOptions(), "key6"));
+  snapshots.push_back(db->GetSnapshot());
+  ASSERT_OK(db->SingleDelete(WriteOptions(), "key7"));
+  snapshots.push_back(db->GetSnapshot());
+  auto* transaction = db->BeginTransaction(WriteOptions());
+  ASSERT_OK(transaction->SetName("txn"));
+  ASSERT_OK(transaction->Put("key1", "value1_2"));
+  ASSERT_OK(transaction->Delete("key2"));
+  ASSERT_OK(transaction->SingleDelete("key3"));
+  ASSERT_OK(transaction->Merge("key4", "value4_2"));
+  ASSERT_OK(transaction->Merge("key5", "value5_3"));
+  ASSERT_OK(transaction->Put("key6", "value6_2"));
+  ASSERT_OK(transaction->Put("key7", "value7_2"));
+  // Prepare but not commit.
+  ASSERT_OK(transaction->Prepare());
+  ASSERT_OK(db->Flush(FlushOptions()));
+  for (auto* s : snapshots) {
+    db->ReleaseSnapshot(s);
+  }
+  // Dummy keys to avoid trivial move.
+  ASSERT_OK(db->Put(WriteOptions(), "a", "dummy"));
+  ASSERT_OK(db->Put(WriteOptions(), "z", "dummy"));
+  ASSERT_OK(db->CompactRange(CompactRangeOptions(), nullptr, nullptr));
+  VerifyInternalKeys({
```

Comment:
Is there a way to make it easier to verify by eyeball checking? Currently it is not clear whether this values actually test the implementation or they are set to make the test pass. For example we can add an item to the verify list right after the final insertion for the key. This would not still help figuring the sequence number.

Reply:
Fixed by adding the expected_seq variable.